rollup/circuitcapacitychecker/impl.go

```diff
@@ -73,46 +73,42 @@ func (ccc *CircuitCapacityChecker) ApplyTransaction(traces *types.BlockTrace) (*
 	}
 
 	encodeStart := time.Now()
-	ccc.jsonBuffer.Reset()
-	err := json.NewEncoder(&ccc.jsonBuffer).Encode(traces)
-	if err != nil {
-		log.Error("fail to json marshal traces in ApplyTransaction", "id", ccc.ID, "TxHash", traces.Transactions[0].TxHash, "err", err)
-		return nil, ErrUnknown
-	}
-
-	tracesStr := C.CString(string(ccc.jsonBuffer.Bytes()))
-	defer func() {
-		C.free(unsafe.Pointer(tracesStr))
-	}()
-
-	rustTrace := C.parse_json_to_rust_trace(tracesStr)
+	rustTrace := MakeRustTrace(traces, &ccc.jsonBuffer)
 	if rustTrace == nil {
 		log.Error("fail to parse json in to rust trace", "id", ccc.ID, "TxHash", traces.Transactions[0].TxHash)
 		return nil, ErrUnknown
 	}
 	encodeTimer.UpdateSince(encodeStart)
 
 	log.Debug("start to check circuit capacity for tx", "id", ccc.ID, "TxHash", traces.Transactions[0].TxHash)
+	return ccc.applyTransactionRustTrace(rustTrace)
+}
+
+func (ccc *CircuitCapacityChecker) ApplyTransactionRustTrace(rustTrace unsafe.Pointer) (*types.RowConsumption, error) {
+	ccc.Lock()
+	defer ccc.Unlock()
+	return ccc.applyTransactionRustTrace(rustTrace)
+}
+
+func (ccc *CircuitCapacityChecker) applyTransactionRustTrace(rustTrace unsafe.Pointer) (*types.RowConsumption, error) {
 	rawResult := C.apply_tx(C.uint64_t(ccc.ID), rustTrace)
 	defer func() {
 		C.free_c_chars(rawResult)
 	}()
-	log.Debug("check circuit capacity for tx done", "id", ccc.ID, "TxHash", traces.Transactions[0].TxHash)
 
 	result := &WrappedRowUsage{}
-	if err = json.Unmarshal([]byte(C.GoString(rawResult)), result); err != nil {
-		log.Error("fail to json unmarshal apply_tx result", "id", ccc.ID, "TxHash", traces.Transactions[0].TxHash, "err", err)
+	if err := json.Unmarshal([]byte(C.GoString(rawResult)), result); err != nil {
+		log.Error("fail to json unmarshal apply_tx result", "id", ccc.ID, "err", err)
```

Comment:
If we ever run into this, it might be very hard to investigate without the txhash. Should we pass it to `applyTransactionRustTrace`?

Reply:
`ApplyTransaction` will have logged it already. Wouldn't that be enough?